src/frontc/cparser.mly

```diff
@@ -244,6 +244,26 @@ let transformOffsetOf (speclist, dtype) member =
       queue;
     Buffer.contents buffer
 
+  (* this makes sure that the labels are only allowed when goto annotation was provided *)
+  let checkAsm asm =
+    let _ = match asm with
+    | ASM (attrs, _, details, _) ->
+      let is_some = function
+        | Some _ -> true
+        | None -> false
+      in
+      let is_none opt = not (is_some opt) in
+      (match details, (List.assoc_opt "goto" attrs) with
+      | Some details, Some _ when is_none details.alabels ->
+        parse_error "expected ':' for labels list in asm goto";
+        raise Parsing.Parse_error
+      | Some details, None when is_some details.alabels ->
+        parse_error "labels provided in inline asm without goto attribute";
+        raise Parsing.Parse_error
+      | _, _ -> ())
+    | _ -> failwith "called checkAsm on non-ASM variant"
+    in asm
+
```

Comment:
```suggestion
  (* this makes sure that the labels are only allowed when goto annotation was provided *)
  let checkAsm asm =
    let _ = match asm with
    | ASM (attrs, _, Some details, _) ->
      (match (List.assoc_opt "goto" attrs) with
      | Some _ when details.alabels = None ->
        parse_error "expected ':' for labels list in asm goto";
        raise Parsing.Parse_error
      | None when details.alabels <> None ->
        parse_error "labels provided in inline asm without goto attribute";
        raise Parsing.Parse_error
      | _ -> ())
    | ASM _ -> ()
    | _ -> failwith "called checkAsm on non-ASM variant"
    in asm

```

Reply:
I'm sorry. I completely forgot that you could use equality on these enums.